src/tests/suite/daos_test_common.c

```diff
@@ -932,35 +932,23 @@ daos_kill_server(test_arg_t *arg, const uuid_t pool_uuid,
 {
 	int		tgts_per_node;
 	int		disable_nodes;
-	int		failures = 0;
-	int		max_failure;
-	int		i;
 	int		rc;
 	char		dmg_cmd[DTS_CFG_MAX];
 
 	tgts_per_node = arg->srv_ntgts / arg->srv_nnodes;
 	disable_nodes = (arg->srv_disabled_ntgts + tgts_per_node - 1) /
 			tgts_per_node;
-	max_failure = (svc->rl_nr - 1) / 2;
-	for (i = 0; i < svc->rl_nr; i++) {
-		if (svc->rl_ranks[i] >=
-		    arg->srv_nnodes - disable_nodes - 1)
-			failures++;
-	}
-
-	if (failures > max_failure) {
-		print_message("Already kill %d targets with %d replica,"
-			      " (max_kill %d) can not kill anymore\n",
-			      arg->srv_disabled_ntgts, svc->rl_nr, max_failure);
-		return;
-	}
 
 	if ((int)rank == -1)
 		rank = arg->srv_nnodes - disable_nodes - 1;
 
+	/* if there is only 1 svc rank, don't kill it */
+	if (svc->rl_nr == 1 && d_rank_in_rank_list(svc, rank)) {
```

Comment:
Do we still only allow (srv_num/2 -1) nodes to be killed? i.e. if more than (srv_num/2 -1 ) nodes are killed, no PS/MS leader will be elected?

Reply:
I'm not sure that I understand the question. If there is only 1 svc rank for the pool, we can't kill it because there are no replicas to take over.

Reply:
I mean if the pool has 5 svcs,  after killing 2, even though there are 3 left, but if kill another one (failure is 3, but quorum is 2 = (5 - 1 )/2), then the new PS leader can not be elected.   So probably we should maintain a svc list (include all svc ranks) in svc to know whether we can kill another svc rank.

Though I just checked all the existing test cases, this check seems enough for all existing test case. So we can improve this later maybe.

Reply:
maybe can see comment from @liw

Reply:
I think the changes made to daos_kill_server is not right. As @wangdi1 said, the proper logic is to check if a majority would be lost. That said, to implement it properly, we need many missing pieces. E.g., because the PS does automatic membership change, after killing a replica, the svc rank list maintained by daos_test becomes stale. We don't have a way to query the latest list of PS replicas though.

@mjmac, would you mind considering reverting the PS replica placement change, please?

Reply:
> @mjmac, would you mind considering reverting the PS replica placement change (#7319), please?

I could do that. Frankly, though, I am concerned that we are missing out on the ability to learn about weaknesses in the product and testing coverage that are due to assumptions being made about the PS placement behavior. The design of DAOS does not dictate that all pool replicas should be on ranks 1-N, does it? If I am correct about the design, then it doesn't seem like we should be ok with tests being implemented that do not allow for full testing of the designed capabilities of the system.

I understand that there are plans to change the placement algorithm based on fault domains, but my feeling is that we are going to run into similar problems down the road once that work starts landing. Since we are now early in the development cycle for 2.2, wouldn't it be better to deal with these problems sooner rather than later? It can be painful, but I think the product will be better in the long run if we get these issues sorted out.

Incidentally, I have another PR (#7409) that adds a new test helper to allow python-based tests to make better choices about which ranks to kill based on the actual pool state rather than a fixed set of assumptions. I believe that the combination of that PR along with a final version of this PR that addresses all requirements should get us back to test stability even with non-deterministic PS replica placement.

Reply:
The point is to do these _before_ or _together_ with the PS replica placement...